File: razer-sdk/src/main/java/net/booone/razersdk/sdk/RzKeyJoiner.java
package net.booone.razersdk.sdk;

import java.util.Collection;
import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;

/**
 * Helper class for joining {@link RzKey} objects and collections. <br><br>
 * Usage example:<br>
 * <pre>
 * Set<RzKey> selectedRzKeys = new RzKeyJoiner()
 *     .with(RzKey.RZKEY_ESC)
 *     .with(Arrays.asList(RzKey.RZKEY_A, RzKey.RZKEY_S))
 *     .with(new RzKeySelector().withRow(0))
 *     .join();
 * //selectedRzKeys contains ESC, A, S and all keys from first top row
 * </pre>
 */
public final class RzKeyJoiner {
    private final EnumSet<RzKey> rzKeys = EnumSet.noneOf(RzKey.class);

    /**
     * Add provided keys to underlying collection.
     *
     * @param array keys to add
     * @return this
     */
    public RzKeyJoiner with(RzKey... array) {
        if (array.length != 0) {
            rzKeys.addAll(Set.of(array));
        }
        return this;
    }

    /**
     * Add all keys from provided collection to underlying collection.
     *
     * @param collection keys to add
     * @return this
     */
    public RzKeyJoiner with(Collection<RzKey> collection) {
        rzKeys.addAll(collection);
        return this;
    }

    /**
     * Add all keys from provided {@link RzKeySelector} to underlying collection.
     *
     * @param rzKeySelector keys to add
     * @return this
     */
    public RzKeyJoiner with(RzKeySelector rzKeySelector) {
        rzKeys.addAll(rzKeySelector.asSet());
        return this;
    }

    /**
     * @return underlying collection
     */
    public Set<RzKey> join() {
        return Collections.unmodifiableSet(rzKeys);
    }
}

File: src/razer-sdk/sdk/RzKeyJoiner.ts
import {RzKey} from './RzKey.js';
import {RzKeySelector} from './RzKeySelector.js';

/**
 * Helper class for joining RzKey objects and collections.
 *
 * Usage example:
 *   const selectedRzKeys = new RzKeyJoiner()
 *       .withKeys(RzKey.RZKEY_ESC)
 *       .withCollection([RzKey.RZKEY_A, RzKey.RZKEY_S])
 *       .withSelector(new RzKeySelector().withRow(0))
 *       .join();
 */
export class RzKeyJoiner {
    private readonly rzKeys = new Set<RzKey>();

    /** Add provided keys to underlying collection. */
    withKeys(...keys: RzKey[]): this {
        for (const key of keys) {
            this.rzKeys.add(key);
        }
        return this;
    }

    /** Add all keys from provided collection to underlying collection. */
    withCollection(collection: Iterable<RzKey>): this {
        for (const key of collection) {
            this.rzKeys.add(key);
        }
        return this;
    }

    /** Add all keys from provided RzKeySelector to underlying collection. */
    withSelector(rzKeySelector: RzKeySelector): this {
        for (const key of rzKeySelector.asSet()) {
            this.rzKeys.add(key);
        }
        return this;
    }

    /** Returns an unmodifiable copy of the underlying collection. */
    join(): ReadonlySet<RzKey> {
        return new Set(this.rzKeys);
    }
}
